File: nildb/secretvault_python/encryption.py
"""Encryption utilities using nilql for secret sharing."""
import base64
import nilql
from typing import List, Tuple, Dict

class CredentialEncryption:
    def __init__(self, num_nodes: int):
        self.num_nodes = num_nodes
        self.secret_key = nilql.SecretKey.generate({'nodes': [{}] * num_nodes},{'store': True})

    def encrypt_password(self, password: str) -> List[str]:
        """Encrypt password using secret sharing."""
        try:
            encrypted_shares = nilql.encrypt(self.secret_key, password)
            print('encrypted_shares', encrypted_shares)
            encoded_shares = []
            
            for i in range(self.num_nodes):
                encoded_shares.append(encrypted_shares[i])

            print('encoded_shares', encoded_shares)
            return encoded_shares
        except Exception as e:
            raise Exception(f"Encryption failed: {str(e)}")

    def decrypt_password(self, encoded_shares: List[str]) -> str:
        """Decrypt password from shares."""
        try:
            decoded_shares = []
            for share in encoded_shares:
                decoded_shares.append(share)
                
            return nilql.decrypt(self.secret_key, decoded_shares)
        except Exception as e:
            raise Exception(f"Decryption failed: {str(e)}")

File: nildb/secretvault_nextjs/app/lib/encryption.ts
// src/lib/encryption.ts
import { nilql } from '@nillion/nilql';

export interface EncryptionConfig {
  nodes: number;
  operations?: {
    store?: boolean;
    match?: boolean;
    sum?: boolean;
  };
}

/**
 * Creates an encryption service for handling credential encryption/decryption
 */
export const createEncryptionService = async (config: EncryptionConfig) => {
  // Create cluster config with specified number of nodes
  const cluster = {
    nodes: Array(config.nodes).fill({}),
  };

  // Initialize secret key with cluster config and operations
  const secretKey = await nilql.ClusterKey.generate(cluster, {
    store: true,
    ...config.operations,
  });

  /**
   * Encrypt a password into shares
   */
  const encryptPassword = async (password: string): Promise<string[]> => {
    try {
      const shares = await nilql.encrypt(secretKey, password);
      if (!Array.isArray(shares)) {
        throw new Error('Unexpected encryption result');
      }
      return shares as string[];
    } catch (error) {
      throw new Error(
        `Encryption failed: ${
          error instanceof Error ? error.message : 'Unknown error'
        }`
      );
    }
  };

  /**
   * Decrypt password from shares
   */
  const decryptPassword = async (shares: string[]): Promise<string> => {
    try {
      if (shares.length !== config.nodes) {
        throw new Error(
          `Expected ${config.nodes} shares but got ${shares.length}`
        );
      }
      const decrypted = await nilql.decrypt(secretKey, shares);
      if (typeof decrypted !== 'string') {
        throw new Error('Unexpected decryption result');
      }
      return decrypted;
    } catch (error) {
      throw new Error(
        `Decryption failed: ${
          error instanceof Error ? error.message : 'Unknown error'
        }`
      );
    }
  };

  return {
    encryptPassword,
    decryptPassword,
  };
};
